### PR TITLE
ScanNet data creation, dataset, and visualization

### PR DESCRIPTION
**Pull request by alexsax**
## Motivation and Context

Adds scannet data set to home_robot, allowing quantitative evaluation of instance maps on real rgbd trajectories

## Changelog

Scannet processing scripts in projects/eval_scannet/scannet/
Scannet dataset in projects/eval_scannet/scannet_dataset.py
Visualization in projects/eval_scannet/visualize_scannet.py
BBox3D utils in src/home_robot/utils/bboxes_3d.py
BBox3D Pyotorch3d visualization in src/home_robot/utils/bboxes_3d_plotly.py

## How Has This Been Tested

Visualized dataset

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: projects/eval_scannet/scannet/scannet_utils.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+"""
+
+import csv
+import os
+
+import numpy as np
+from plyfile import PlyData
+
+
+def represents_int(s):
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/scannet_utils.py:17]: general utilities like htis could be moved to src/home_robot?

File: projects/eval_scannet/scannet/scannet_utils.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        bool: Whether s represents int or not.
+    """
+    try:
+        int(s)
+        return True
+    except ValueError:
+        return False
+
+
+def read_label_mapping(filename, label_from="raw_category", label_to="nyu40id"):
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/scannet_utils.py:33]: please add docstring + type hints for all new code

File: projects/eval_scannet/scannet/extract_posed_images.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
+# Modified from
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/extract_posed_images.py:1]: license header at top of file
+ we should move to src/home_robot/home_robot/datasets/scannet.py i think

File: projects/eval_scannet/scannet/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
+### Prepare ScanNet Data for Indoor Detection or Segmentation Task
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/README.md:1]: Why is the readme not in projects/eval_scannet?

File: projects/eval_scannet/scannet/scannet_utils.py (hunk 1 above)
Comment (cpaxton) [reply projects/eval_scannet/scannet/scannet_utils.py:17]: i think because this is specifically for the dataset we could also move it to src/home_robot/home_robot/datasets/scannet_utils.py?

File: projects/eval_scannet/scannet/load_scannet_data.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+import json
+import os
+
+import numpy as np
+import scannet_utils
+
+currentdir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+
+
+def read_aggregation(filename):
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/load_scannet_data.py:20]: docstring please

File: projects/eval_scannet/scannet/load_scannet_data.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            object_id_to_segs[object_id] = segs
+            if label in label_to_segs:
+                label_to_segs[label].extend(segs)
+            else:
+                label_to_segs[label] = segs
+    return object_id_to_segs, label_to_segs
+
+
+def read_segmentation(filename):
+    assert os.path.isfile(filename)
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/load_scannet_data.py:42]: pathlib is a bit better
file = Path(filename)
file.exists()
with file.open() as f

Comment (cpaxton) [reply projects/eval_scannet/scannet/load_scannet_data.py:42]: from pathlib import Path

File: projects/eval_scannet/scannet/load_scannet_data.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        for i in range(num_verts):
+            seg_id = data["segIndices"][i]
+            if seg_id in seg_to_verts:
+                seg_to_verts[seg_id].append(i)
+            else:
+                seg_to_verts[seg_id] = [i]
+    return seg_to_verts, num_verts
+
+
+def extract_bbox(mesh_vertices, object_id_to_segs, object_id_to_label_id, instance_ids):
```
Comment (cpaxton) [inline projects/eval_scannet/scannet/load_scannet_data.py:56]: gets per-instance bounding boxes from the mesh vertices associated with each object, from the scannet labels?

File: projects/eval_scannet/.gitignore [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Files not ready for commit
+test_bboxes_3d.py
```
Comment (cpaxton) [inline projects/eval_scannet/.gitignore:2]: do these work if they're in a subfolder like this?

File: src/home_robot/home_robot/utils/bboxes_3d.py [hunk 9]
```diff
@@ -0,0 +1,1126 @@
...
+
+    def __init__(
+        self,
+        bounds: List[torch.Tensor],
+        features: Optional[List[torch.Tensor]] = None,
+        names: Optional[List[torch.Tensor]] = None,
+    ) -> None:
+        """
+        Args:
+            bounds:
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/bboxes_3d.py:41]: Very nice documentation

File: src/home_robot/home_robot/utils/bboxes_3d.py [hunk 10]
```diff
@@ -0,0 +1,1126 @@
...
+                - List where each element is a tensor of shape (num_boxes, 3)
+                  containing the normal vector for each point.
+                Not Implemented: - Padded float tensor of shape (num_scenes, num_boxes, 3).
+            features:
+                Can be either
+
+                - None
+                - List where each element is a tensor of shape (num_boxes, C)
+                  containing the features for the boxes in the scene.
+                Not Implemented: - Padded float tensor of shape (num_scenes, num_boxes, C).
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/bboxes_3d.py:60]: should we check for this case and raise exception?

File: src/home_robot/home_robot/utils/bboxes_3d.py [hunk 11]
```diff
@@ -0,0 +1,1126 @@
...
+            break
+
+        # keep the boxes with IoU less than thresh_iou
+        # _, iou = box3d_overlap(bounding_boxes[idx].unsqueeze(0), bounding_boxes[order])
+        mask = iou[idx, order] < iou_threshold
+        order = order[mask]
+    return torch.tensor(keep), vol, iou
+
+
+# def get_cuboid_verts_faces(box3d=None, R=None):
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/bboxes_3d.py:1023]: is this just dead code?

**Review by cpaxton — APPROVED** [review]
LGTM - minor comments but nothing blocking

**Comment by cpaxton** [comment]
Actually we should probably do the refactor suggested before merge: move things into src/home_robot + add license headers and comments

**Comment by alexsax** [comment]
Sounds good -- I'll do the refactor + address the changes (licenses, explaining what the code is used for, etc) and add a new commit. Then merge :)

File: projects/eval_scannet/.gitignore (hunk 8 above)
Comment (alexsax) [reply projects/eval_scannet/.gitignore:2]: Yeah :) it's pretty useful for keeping them clean + porting across projects

File: projects/eval_scannet/scannet/load_scannet_data.py (hunk 7 above)
Comment (alexsax) [reply projects/eval_scannet/scannet/load_scannet_data.py:56]: Yes -- in practice the mesh vertices here are downsampled a bit from the full mesh (50k points/scene), following the convention of VoteNet and MMDet3d

File: projects/eval_scannet/scannet/load_scannet_data.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        label_id = object_id_to_label_id[obj_id]
+        obj_pc = mesh_vertices[instance_ids == obj_id, 0:3]
+        if len(obj_pc) == 0:
+            continue
+        xyz_min = np.min(obj_pc, axis=0)
+        xyz_max = np.max(obj_pc, axis=0)
+        bbox = np.concatenate(
+            [(xyz_min + xyz_max) / 2.0, xyz_max - xyz_min, np.array([label_id])]
+        )
+        # NOTE: this assumes obj_id is in 1,2,3,.,,,.NUM_INSTANCES
```
[the remark this answers is not in the document]
Comment (alexsax) [reply projects/eval_scannet/scannet/load_scannet_data.py:69]: It's true for the scannet code, which is all this script is supposed to do :)

**Review by cpaxton — COMMENTED** [review]
Do we need an __init__.py in datasets and scannet and data?
